### PR TITLE
Added Unix Implementations for omrsock_connect and omrsock_accept functions

### PR DESCRIPTION
**Pull request by caohaley**
- Added `omrsock_connect` and `omrsock_accept` implementations which simply calls the system socket functions and returns appropriate error codes other than `omrsock_accept` which also allocates an OMRSocket structure containing the returned new socket for client server communications.
- Added test implementations for the two functions which attempts to set up a connection between server and client for stream, and tries to set up peer address for datagram

Issue: #5043

### CODE REVIEW

File: port/unix/omrsock.c [hunk 1]
```diff
@@ -490,13 +490,58 @@ omrsock_listen(struct OMRPortLibrary *portLibrary, omrsock_socket_t sock, int32_
...
+		return OMRPORT_ERROR_SOCK_CONNECT_FAILED;
+	}
+	return 0;
 }
 
 int32_t
 omrsock_accept(struct OMRPortLibrary *portLibrary, omrsock_socket_t serverSock, omrsock_sockaddr_t addrHandle, omrsock_socket_t *sockHandle)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	int32_t	 connSocketDescriptor;
```
Comment (rwy0717) [inline port/unix/omrsock.c:516]:
```suggestion
	int32_t connSocketDescriptor;
```

File: fvtest/porttest/omrsockTest.cpp [hunk 2]
```diff
@@ -437,21 +474,83 @@ TEST(PortSockTest, create_IPv4_mapped_IPv6_Socket_Address)
...
+TEST(PortSockTest, two_socket_stream_communication)
+{
+	OMRPORT_ACCESS_FROM_OMRPORT(portTestEnv->getPortLibrary());
+
+	OMRSockAddrStorage serverStreamSockAddr;
+	omrsock_socket_t serverStreamSocket = NULL;
+	uint16_t port = 4930;
+	uint8_t serverAddr[4];
+
+	// To Create a Server Socket and Address
```
Comment (rwy0717) [inline fvtest/porttest/omrsockTest.cpp:497]: We tend to use `/* blah */` style comments, rather than these (imho better) single-line comments.

Comment (caohaley) [reply fvtest/porttest/omrsockTest.cpp:497]: Sounds good! I will change it.

File: port/unix/omrsock.c [hunk 3]
```diff
@@ -490,13 +494,58 @@ omrsock_listen(struct OMRPortLibrary *portLibrary, omrsock_socket_t sock, int32_
...
 omrsock_connect(struct OMRPortLibrary *portLibrary, omrsock_socket_t sock, omrsock_sockaddr_t addr)
 {
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	socklen_t addrLength;
+
+	if (NULL == addr || NULL == sock) {
+		return OMRPORT_ERROR_INVALID_ARGUMENTS;
+	}
+
+	if (OS_SOCK_AF_INET == (addr->data).ss_family) {
```
Comment (rwy0717) [inline port/unix/omrsock.c:503]:
```suggestion
	if (OS_SOCK_AF_INET == addr->data.ss_family) {
```

File: port/unix/omrsock.c [hunk 4]
```diff
@@ -560,15 +609,32 @@ omrsock_htonl(struct OMRPortLibrary *portLibrary, uint32_t val)
 int32_t
 omrsock_inet_pton(struct OMRPortLibrary *portLibrary, int32_t addrFamily, const char *addr, uint8_t *addrNetworkOrder)
```
Comment (rwy0717) [inline port/unix/omrsock.c:610]: I think you should rename `addrNetworkOrder` to `result`.

File: port/unix/omrsock.c [hunk 5]
```diff
@@ -560,15 +609,32 @@ omrsock_htonl(struct OMRPortLibrary *portLibrary, uint32_t val)
...
 	}
 
-	if (1 != inet_pton(get_os_family(addrFamily), addr, addrNetworkOrder)) {
+#if defined(J9ZOS390) && !defined(OMR_EBCDIC)
+	addrA2e = a2e(addr, strlen(addr));
+	rc = inet_pton(get_os_family(addrFamily), addrA2e, (void *)addrNetworkOrder);
+	free(addrA2e);
+#else /* defined(J9ZOS390) && !defined(OMR_EBCDIC) */
+	rc = inet_pton(get_os_family(addrFamily), addr, (void *)addrNetworkOrder);
+#endif /* defined(J9ZOS390) && !defined(OMR_EBCDIC) */
```
Comment (rwy0717) [inline port/unix/omrsock.c:627]: Maybe simpler as:

```c++
#if defined(J9ZOS390) && !defined(OMR_EBCDIC)
	addr = a2e(addr, strlen(addr));
#endif

	rc = inet_pton(get_os_family(addrFamily), addr, (void *)result);

#if defined(J9ZOS390) && !defined(OMR_EBCDIC)
	free(addr);
#endif
```

and then you can get rid of the `addrA2e` local as well.

Comment (caohaley) [reply port/unix/omrsock.c:627]: Sounds good, I will change it.

Comment (caohaley) [reply port/unix/omrsock.c:627]: @rwy0717 There is just one problem with changing it, addr is a const char * right now. I could change the addr argument to be a char *, but that may not be a good idea.

Is it okay if I do:
```
#if defined(J9ZOS390) && !defined(OMR_EBCDIC)
	char *addrIn = a2e(addr, strlen(addr));
#else /* defined(J9ZOS390) && !defined(OMR_EBCDIC) */
	const char *addrIn = addr;
#endif /* defined(J9ZOS390) && !defined(OMR_EBCDIC) */

	rc = inet_pton(get_os_family(addrFamily), addrIn, (void *)addrNetworkOrder);

#if defined(J9ZOS390) && !defined(OMR_EBCDIC)
	free(addrIn);
#endif /* defined(J9ZOS390) && !defined(OMR_EBCDIC) */
```

Comment (rwy0717) [reply port/unix/omrsock.c:627]: Oh you're right, I'm fine with the original then.

**Comment by caohaley** [comment]
@rwy0717 Ready for another round of review.

**Comment by caohaley** [comment]
@rwy0717 Ready for another round of review.

File: port/unix/omrsock.c [hunk 6]
```diff
@@ -490,13 +494,58 @@ omrsock_listen(struct OMRPortLibrary *portLibrary, omrsock_socket_t sock, int32_
...
+	connSocketDescriptor = accept(serverSock->data, (omr_os_sockaddr *)&addrHandle->data, &addrLength);
+	if (connSocketDescriptor < 0) {
+		portLibrary->error_set_last_error(portLibrary, errno, OMRPORT_ERROR_SOCK_ACCEPT_FAILED);
+		return OMRPORT_ERROR_SOCK_ACCEPT_FAILED;
+	}
+
+	*sockHandle = (omrsock_socket_t)portLibrary->mem_allocate_memory(portLibrary, sizeof(struct OMRSocket), OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+	if (*sockHandle == NULL) {
+		close(connSocketDescriptor);
+		*sockHandle = NULL;
```
Comment (rwy0717) [inline port/unix/omrsock.c:543]: We need to free the socket memory on error.

Comment (rwy0717) [inline port/unix/omrsock.c:543]: You don't need to set `*sockHandle` to null, because it is already null.

**Comment by rwy0717** [comment]
@genie-omr build all

**Pull request merged**